src/EditorFeatures/TestUtilities/NavigateTo/AbstractNavigateToTests.cs

```diff
@@ -248,14 +258,39 @@ public FirstDocIsActiveAndVisibleDocumentTrackingService(Workspace workspace)
             _workspace = workspace;
         }
 
-        public event EventHandler<DocumentId> ActiveDocumentChanged;
-        public event EventHandler<EventArgs> NonRoslynBufferTextChanged;
+        public event EventHandler<DocumentId> ActiveDocumentChanged { add { } remove { } }
+        public event EventHandler<EventArgs> NonRoslynBufferTextChanged { add { } remove { } }
 
         public DocumentId GetActiveDocument()
             => _workspace.CurrentSolution.Projects.First().DocumentIds.First();
 
         public ImmutableArray<DocumentId> GetVisibleDocuments()
             => ImmutableArray.Create(_workspace.CurrentSolution.Projects.First().DocumentIds.First());
     }
-#pragma warning restore
+
+    [Export]
+    [ExportWorkspaceServiceFactory(typeof(IDocumentTrackingService), ServiceLayer.Host)]
```

Comment:
first export seems redundant.

Reply:
➡️ It's used here to obtain the singleton outside of the context of a workspace:

https://github.com/sharwell/roslyn/blob/42143b89e184b605dc6f5e0413a0b99c53119ba5/src/EditorFeatures/TestUtilities/NavigateTo/AbstractNavigateToTests.cs#L114

It's a "common pattern" for singleton factories that get accessed like this, but could still be added to the documentation.

Reply:
ah, ok :)